File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/Greedy/JobSequencing.ts
class Job {
	id: string;
	deadline: number;
	profit: number;

	constructor(id: string, deadline: number, profit: number) {
		this.id = id;
		this.deadline = deadline;
		this.profit = profit;
	}
}

class JobSequencing {
	jobs: Job[];
	n: number;
	maxDL: number;

	constructor(ids: string[], deadlines: number[], profits: number[], n: number) {
		this.jobs = new Array<Job>(n);
		this.n = n;
		this.maxDL = Math.max(...deadlines);

		for (let i = 0; i < n; i++) {
			this.jobs[i] = new Job(ids[i], deadlines[i], profits[i]);
		}
	}

	print(): void {
		this.jobs.sort((a, b) => b.profit - a.profit);
		const result: boolean[] = new Array<boolean>(this.maxDL).fill(false);
		const job: string[] = new Array<string>(this.maxDL);
		let profit = 0;

		for (let i = 0; i < this.n; i++) {
			for (let j = this.jobs[i].deadline - 1; j >= 0; j--) {
				if (!result[j]) {
					result[j] = true;
					job[j] = this.jobs[i].id;
					profit += this.jobs[i].profit;
					break;
				}
			}
		}

		console.log("Profit is:", profit);
		let output = "Jobs selected are:";
		for (let i = 0; i < this.maxDL; i++) {
			if (job[i] !== undefined) {
				output += " " + job[i];
			}
		}
		console.log(output);
	}
}

const ids: string[] = ['a', 'b', 'c', 'd', 'e'];
const deadlines: number[] = [3, 1, 2, 4, 4];
const profits: number[] = [50, 40, 27, 31, 30];
const jobSequencing: JobSequencing = new JobSequencing(ids, deadlines, profits, 5);
jobSequencing.print();

/*
Profit is: 151
Jobs selected are: b e a d
*/

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/Greedy/JobSequencing.swift
import Foundation

class Job : Comparable {
    var id : Character
    var deadline : Int
    var profit : Int

    init(_ id : Character, _ deadline : Int, _ profit : Int) {
        self.id = id
        self.deadline = deadline
        self.profit = profit
    }

    static func < (lhs: Job, rhs: Job) -> Bool {
        return lhs.profit < rhs.profit
    }

    static func > (lhs: Job, rhs: Job) -> Bool {
        return lhs.profit > rhs.profit
    }

    static func == (lhs: Job, rhs: Job) -> Bool {
        return lhs.profit == rhs.profit
    }
}

func JobSequencing(_ ids : inout [Character], _ deadlines : inout [Int], _ profits : inout [Int], _ n : Int) {
    var maxDL = deadlines[0]
    var i : Int = 1
    while (i < n) {
        maxDL = max(maxDL, deadlines[i])
        i += 1
    }
    
    var jobs : [Job] = []
    i = 0
    while (i < n) {
        jobs.append(Job(ids[i], deadlines[i], profits[i]))
        i += 1
    }
    
    jobs.sort(by: >)// Decreasing order.
    
    var result : [Bool] = Array(repeating: false, count: maxDL)
    var answer : [Character] = Array(repeating: "0", count: maxDL)
    var profit : Int = 0

    i = 0 // Iterate through all given jobs
    while (i < n) {
        var j : Int = jobs[i].deadline - 1
        while (j >= 0) {
            if (result[j] == false) {
                result[j] = true
                answer[j] = jobs[i].id
                profit += jobs[i].profit
                break
            }
            j -= 1
        }
        i += 1
    }
    
    print("Profit is :: " + String(profit))
    print("Jobs selected are ::", terminator: " ")
    i = 0
    while (i < maxDL) {
        if (answer[i] != "0") {
            print(String(answer[i]), terminator: " ")
        }
        i += 1
    }
}

// Testing code.
var id : [Character] = ["a", "b", "c", "d", "e"]
var deadline : [Int] = [3, 1, 2, 4, 4]
var profit : [Int] = [50, 40, 27, 31, 30]
JobSequencing(&id, &deadline, &profit, 5)
/* 
Profit is :: 151
Jobs selected are :: b e a d 
 */